meteor/sofie-server/business-logic/services/superfly-timeline-builder.ts

```diff
@@ -0,0 +1,435 @@
+			id: `${ACTIVE_GROUP_PREFIX}${activePart.id}`,
+			priority: HIGH_PRIORITY,
+			isGroup: true,
+			children: [],
+			enable: currentPartEnable,
+			layer: '',
+			shouldAutoNextAtPointInTime,
+			content: {
+				type: TimelineObjectType.GROUP,
+				deviceType: DeviceType.ABSTRACT,
+			},
+		}
+
+		activeGroup.children = activePart
+			.getPieces()
+			.flatMap((piece) => this.generateGroupsAndTimelineObjectsForPiece(piece, activePart, activeGroup))
+		return activeGroup
+	}
+
+	private generateGroupsAndTimelineObjectsForPiece(
+		piece: Piece,
+		part: Part,
+		parentGroup: TimelineObjectGroup
+	): TimelineObject[] {
+		const timelineObjectsToReturn: TimelineObject[] = []
+		const pieceEnable: TimelineEnable | undefined = this.createTimelineEnableForPiece(part, piece, parentGroup)
+
+		if (!pieceEnable) {
+			return []
+		}
+
+		if (pieceEnable.start === 'now') {
+			// TODO: We would like to try to not use "now", but they might be necessary for AdLibs. We have to double check when implementing AdLibs.
+			// TODO: This is mainly to alert of if we ever find a Piece with start==="now" while implementing the Timeline. Should be removed when done.
+			throw new UnsupportedOperation(
+				`Found an enable.start="start" for control for Piece: ${piece.id}. We are trying to avoid those if possible.`
+			)
+		}
+
+		const controlForPiece: TimelineObjectControl = this.createTimelineObjectControl(parentGroup, piece, pieceEnable)
+		timelineObjectsToReturn.push(controlForPiece)
+
+		const childGroupForPiece: TimelineObjectGroup = this.createTimelineObjectGroupForPiece(
+			parentGroup,
+			piece,
+			controlForPiece
+		)
+		timelineObjectsToReturn.push(childGroupForPiece)
+
+		if (this.shouldPieceHavePreRollGroup(controlForPiece, piece)) {
+			const preRollControlForPiece: TimelineObjectControl = this.createPreRollGroupForControl(
+				controlForPiece,
+				parentGroup
+			)
+			timelineObjectsToReturn.push(preRollControlForPiece)
+
+			controlForPiece.enable.start = `#${preRollControlForPiece.id} + ${piece.preRollDuration}`
+		}
+
+		// TODO: Core resolves Endcaps, but endCaps seems to be related to when a PieceInstance was stopped which is a concept we don't use.
+		// TODO: EnaCaps seems to be used to check when infinite pieces should stop. That logic should be in Rundown.ts so we shouldn't need to worry about that here
+		// TODO: EndCaps also seems to be used if the Part/Piece needs to stop i.e. a Server will end at some point or an overlay "bundt" is only shown for x seconds. But Blueprint define that on ingest on the TimelineObjects?
+		// TODO: Trying to leave out EndCaps for now and see how the Rundown behaves. Downside? We might have issues with "pieceStartOffset", whatever that is.
+
+		childGroupForPiece.children = piece.timelineObjects.map((timelineObject) =>
+			this.mapToTimelineObjectForPieceGroup(timelineObject, childGroupForPiece, piece)
+		)
+
+		return timelineObjectsToReturn
+	}
+
+	private createTimelineEnableForPiece(
```

Comment:
I have always been told and felt that the golden lenght of a method is less than 20 lines. It can be okay up to 30, but thats should be the limit. 

If a method goes above that, then something in the method should propably be pulled out to a sub-method(s).
Don't know if it is critical in this case, but one could pull the body of each case, out into their own methods.
That would significantly shrink this method, and lowers it respnsibilities.
Something along the lines of the following could be used.

```typescript
...
switch (piece.transitionType) {
   case TransitionType.IN_TRANSITION: return handleTransitionTypeIn(...)
   case TransitionType.OUT_TRANSITION: return handleTransitionTypeOut(...)
   case TransitionType.NO_TRANSITION: return handleTransitionTypeNone(...)
   default: {
      ...
   }
}
...
```

Reply:
I think we have all heard a different number for the "golden rule".
It's important to note it's a rule of thumb and not set in stone. Methods are allowed to go over that line number _when it makes sense_.

That being said, I agree, it would make sense here to extract each case to a separate method

Reply:
Just a small note to your naming suggestion. We are creating `TimelineEnable` so a more accruate name for the methods does would be `createInTransitionTimelineEnable()` etc.

The naming convention `handleX` doesn't really indicate what the method does.

Reply:
The names was also more as an indication for what piece of code should go where.
Happy to hear that you got a better name, and the cases are being moved to seperate methods.